[IIIF-900] Update build to include Docker and Kakadu

ksclarke opened the pull request:
_**Note:** There are conflicts between my branch and David's. We can wait to review mine until David's is in and I've done the conflict resolution._

**IIIF-900**
* Rewrite the build to do the Docker build as a part of this project's build; this eliminates the need for the docker-bucketeer project and the CodeBuild setup in AWS.
* Rewrite Travis to support the new build; it will now publish Docker images to DockerHub too.
* Add the Kakadu build to Travis now too; it will publish a Docker image to a private DockerHub repo.
* Update documentation/README.
* Simplify Dockerfile (but this also means it no longer stands apart from our Maven build).
* Add UTF-8 OS support to our Docker image (seems we didn't have a locale set before).
* Update dependencies (for security issues) and parent project.
* Update Javadocs to conform to stricter Checkstyle rules from the parent project.
* Update logging so the LEVEL can be set dynamically (the same way Fester's is).
* Add GitHub encypted key so Travis can pull Kakadu from our private repo at test time.
* Remove the old IT scripts... yay!

**IIIF-944** Bucketeer tests timing out on Travis
* Add an async handler to the `.listen()` call on app startup. It seems on some places, like Travis, it wasn't able to grab the port and start the application by the time the tests run. This makes sure the app startup doesn't proceed until it has that port and the app is listening on it.
* Add a check on the test startup so it will check to see if the port it's testing against has been assigned, and if it hasn't it outputs a warning message in the logs. 

**IIIF-945** Fix Bucketeer response handlers + tests
* Change the way the router handles static pages.
* Wrote MatchingOpNotFoundHandler to handle instances when the OpenAPI configured router doesn't catch something (and we want to provide a more detailed response than just a 500 error).
* Improve various tests, including in particular ImageUpload(Kakadu)IT.
* Add some additional exception classes needed by test improvements.
* Move Kakadu tests into an IT that only runs against the Docker container (built with Kakadu).
* Improve the way the presence of Kakadu is detected and removed some workarounds in tests re: this.
* Update uses of deprecated Vertx APIs to the current forms (just some, not all... just the ones from the classes I was touching).
* Fix runtime detection of Linux system, which will allow optimizations in Netty (the underlying server beneath Vertx).
* Update OpenAPI endpoint `/{image-id}/{file-path}` to `/images/{image-id}/{file-path}` to make routing easier (this is a breaking change, but noone is really using this path... yet... we're getting ready to).

ksclarke committed:
[IIIF-893] Update build; added Docker build

ksclarke committed:
Fix image load path in IT

DRickard commented:
You seem to have duped entries for the Kakadu repo in pom.xml (lines 77-81):

```
    <!-- Git repo with Kakadu source code (ours is private; override with yours) -->
    <kakadu.git.repo>scm:git:git@github.com:uclalibrary/kakadu.git</kakadu.git.repo>

    <!-- Git repo with Kakadu source code (ours is private; override with yours) -->
    <kakadu.git.repo>scm:git:git@github.com:uclalibrary/kakadu.git</kakadu.git.repo>

```

ksclarke commented:
> You seem to have duped entries for the Kakadu repo in pom.xml (lines 77-81):
> 
> ```
>     <!-- Git repo with Kakadu source code (ours is private; override with yours) -->
>     <kakadu.git.repo>scm:git:git@github.com:uclalibrary/kakadu.git</kakadu.git.repo>
> 
>     <!-- Git repo with Kakadu source code (ours is private; override with yours) -->
>     <kakadu.git.repo>scm:git:git@github.com:uclalibrary/kakadu.git</kakadu.git.repo>
> ```

Doh! Thanks, removing the second one :-)

ksclarke committed:
Fix POM config/property duplication

DRickard commented:
Minor nit, but in FesterVerticle.java, line 94:
`message.fail(500, throwable.getMessage());`
why not use `HTTP.INTERNAL_SERVER_ERROR`?

ksclarke committed:
Fix stray logging/error handling

markmatney reviewed (changes requested):
I found a few minor issues :wink: interesting learning about the platform-specific Netty optimizations.

ksclarke commented on src/main/docker/Dockerfile:
```diff
@@ -0,0 +1,102 @@
+EXPOSE 8888
+
+# Update packages and install tools
+#  Removing /var/lib/apt/lists/* prevents using `apt` unless you do `apt update` first
```
Reduces image size and conveys the idea that this isn't a thing that's intended to be updated. New images should be minted.

markmatney replied:
cool, I figured it was a guard

ksclarke commented on pom.xml:
```diff
@@ -47,39 +47,80 @@
       <organizationUrl>http://github.com/uclalibrary</organizationUrl>
+      <timezone>America/Los_Angeles</timezone>
+    </developer>
+    <developer>
+      <id>markmatney</id>
+      <name>Mark A. Matney, Jr.</name>
+      <email>mmatney@library.ucla.edu</email>
+      <organization>UCLA Library</organization>
+      <organizationUrl>http://github.com/uclalibrary</organizationUrl>
+      <timezone>America/Los_Angeles</timezone>
+    </developer>
```
You'll be working on it  ;-)

markmatney replied:
LOL :crystal_ball:

ksclarke commented on src/main/java/edu/ucla/library/bucketeer/handlers/MatchingOpNotFoundHandler.java:
```diff
@@ -0,0 +1,49 @@
+/**
+ * A handler for requests that do not match the operations defined in the Bucketeer OpenAPI specification.
+ * <p>
+ * This handler can fail the context with any response code other than a 404. Because it's set up to catch 404s, failing
+ * the routing context with a 404 will create a loop. To return a legitimate 404 from this handler, use
+ * HttpServerReponse and set 404 as the response code, remembering to call <code>.end()</code> on the response.
```
It's more along the lines of documentation for the future, based on past troubles (this class is actually copied from Fester where I learned this lesson the hard way; I debated removing it but thought my future self would thank me for leaving it).

ksclarke commented on src/main/resources/bucketeer_messages.xml:
```diff
@@ -171,4 +172,11 @@
   <entry key="BUCKETEER-500">{}</entry>
   <entry key="BUCKETEER-501">Large image ({} [{}]) skipped; maximum allowed size set for: {}</entry>
+  <entry key="BUCKETEER-502">Bucketeer is now listening on port: {}</entry>
+  <entry key="BUCKETEER-503">Resource not found: {}</entry>
```
I must have changed my mind about it. Removing the entry from the XML file.

ksclarke committed:
Update with PR review suggestions

ksclarke committed:
Add workaround for JDK 11 bug that's showed up in Travis 2x now

Cf. https://github.com/slackapi/java-slack-sdk/issues/465#issuecomment-632150000
